src/Features/CSharp/Portable/UpgradeProject/CSharpUpgradeProjectCodeFixProvider.cs

```diff
@@ -92,7 +92,7 @@ public override bool IsUpgrade(Project project, string newVersion)
         {
             Contract.ThrowIfFalse(LanguageVersionFacts.TryParse(newVersion, out var parsedNewVersion));
 
-            var parseOptions = (CSharpParseOptions)project.ParseOptions;
+            var parseOptions = (CSharpParseOptions)project.ParseOptions!;
```

Comment:
Can parseOptions be null?

Reply:
Generally yes if it's a F# or TypeScript project. But not for C# projects.

Reply:
What about adding a `Contract.ThrowIfNull` at the start of this method? I believe that should eliminate the need to use `!`.

Reply:
Yes, that's a good suggestion.

Reply:
I still really hate '!' at the end of a line... 

Does this reflect that the design is too complicated or wrong, or do we expect this to happen a lot?

Reply:
Maybe we can add a `Contract.ThrowIfNotCSharp` where we have a bunch of null checks that flow analysis can then use as hints... is that too expensive?

Reply:
What sort of API shape did you imagine? There's no way to have an API like: `void ThrowIfNotCSharp(Project project) { … }` that then annotates that project.ParseOptions will be not-null. The only thing would be to directly just say Contract.ThrowIfNull(project.ParseOptions) which is just good enough here.

Reply:
That's disappointing. I forgot we can't annotate that a specific property is not null, and flow analysis doesn't work across functions :(